ethcore/src/executive.rs

```diff
@@ -34,14 +34,29 @@ pub use types::executed::{Executed, ExecutionResult};
 /// Maybe something like here: `https://github.com/ethereum/libethereum/blob/4db169b8504f2b87f7d5a481819cfb959fc65f6c/libethereum/ExtVM.cpp`
 const STACK_SIZE_PER_DEPTH: usize = 24*1024;
 
-/// Returns new address created from address and given nonce.
-pub fn contract_address(address: &Address, nonce: &U256) -> Address {
+/// Returns new address created from address, nonce, and code hash
+pub fn contract_address(address_scheme: CreateContractAddress, sender: &Address, nonce: &U256, code_hash: &H256) -> Address {
 	use rlp::RlpStream;
 
-	let mut stream = RlpStream::new_list(2);
-	stream.append(address);
-	stream.append(nonce);
-	From::from(stream.out().sha3())
+	match address_scheme {
+		CreateContractAddress::FromSenderAndNonce => {
+			let mut stream = RlpStream::new_list(2);
+			stream.append(sender);
+			stream.append(nonce);
+			From::from(stream.as_raw().sha3())
+		},
+		CreateContractAddress::FromCodeHash => {
+			let mut buffer = [0xffu8; 20 + 32];
```

Comment:
I think the fix is wrong.

The relevant [EIP](https://github.com/ethereum/EIPs/pull/208/files#diff-95b306e6e6d2d6dbb0c5fd3ac4ea45c4R26) says the pad is 0xffff..ff.  The NULL_SENDER is [defined](https://github.com/ethereum/EIPs/pull/208/files#diff-95b306e6e6d2d6dbb0c5fd3ac4ea45c4R19) to be 0xffff..ff.